### PR TITLE
Limit the number of valid hole fits

### PR DESCRIPTION
**Pull request by expipiplus1**
Originally https://github.com/haskell/ghcide/pull/888

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/1914"></a>

### CODE REVIEW

**Comment by jneira** [comment]
@pepeiborra commented in the original pr

> I think the limit (be it 5 or 15) should be moved to the GHC flags that we set in GHC.Compat, and make sure we don't override a user value.

I think we should take in account the ghc flag insome way, at least to honour it if the user has set it on purpose.
I would be a little bit surprised if i set 20 and only see 10 in the editor.

File: ghcide/src/Development/IDE/GHC/Compat.hs [hunk 1]
```diff
@@ -303,7 +304,7 @@ setUpTypedHoles df
   $ df
   { refLevelHoleFits = Just 1   -- becomes slow at higher levels
   , maxRefHoleFits   = Just 10  -- quantity does not impact speed
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/GHC/Compat.hs:306]: Let's apply the same trick to all the other flags
```suggestion
  { refLevelHoleFits = refLevelHoleFits df <|> Just 1   -- becomes slow at higher levels
  , maxRefHoleFits   = maxRefHoleFits df <|> Just 10  -- quantity does not impact speed
```

**Review by pepeiborra — APPROVED** [review]

**Comment by pepeiborra** [comment]
Thanks Joe!

**Comment by Ailrun** [comment]
@expipiplus1 Tests from
https://github.com/haskell/haskell-language-server/blob/b970e250217d970060a211022852d863e34060aa/ghcide/test/exe/Main.hs#L2433-L2458
uses a filtered action, so these keep failing.

**Comment by expipiplus1** [comment]
hmm, the tests don't run for me locally. Would it be possible for someone else to take over this PR, please?

```
❯ cabal run test:ghcide-tests -- --pattern "postfix hole"
Up to date
ghcide
  code actions
    fill typed holes
      postfix hole uses postfix notation of infix operator: FAIL
        Exception: ghcide: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
```

**Comment by isovector** [comment]
Your CI appears to be blocked by https://github.com/haskell/haskell-language-server/issues/2016

**Comment by jneira** [comment]
tests failing:

```
 Test suite ghcide-tests: RUNNING...
ghcide
  code actions
    fill typed holes
      postfix hole uses postfix notation of infix operator: FAIL (1.82s)
        test/exe/Main.hs:5546:
        Found no matching actions for "replace _ with (+)" in ["replace _ with test","replace _ with (-)","replace _ with (*)","replace _ with asTypeOf","replace _ with const","replace _ with pure","replace _ with (Some hole fits suppressed; use -fmax-valid-hole-fits=N or -fno-max-valid-hole-fits)","replace _ with curry _","replace _ with ($) _","replace _ with ($!) _","replace _ with const _","replace _ with flip _","replace _ with id _"]
        Use -p '/postfix hole uses postfix notation of infix operator/' to rerun this test only.
      filling infix type hole uses infix operator:          FAIL (1.83s)
        test/exe/Main.hs:5546:
        Found no matching actions for "replace _ with (+)" in ["replace _ with test","replace _ with (-)","replace _ with (*)","replace _ with asTypeOf","replace _ with const","replace _ with pure","replace _ with (Some hole fits suppressed; use -fmax-valid-hole-fits=N or -fno-max-valid-hole-fits)","replace _ with curry _","replace _ with ($) _","replace _ with ($!) _","replace _ with const _","replace _ with flip _","replace _ with id _"]
        Use -p '/filling infix type hole uses infix operator/' to rerun this test only.
  cradle
    dependencies
```

**Comment by jneira** [comment]
```
ghcide
  code actions
    fill typed holes
      postfix hole uses postfix notation of infix operator: FAIL (1.72s)
        test/exe/Main.hs:5547:
        Found no matching actions for "replace _ with (+)" in ["replace _ with test","replace _ with (-)","replace _ with (*)","replace _ with asTypeOf","replace _ with const","replace _ with pure","replace _ with (Some hole fits suppressed; use -fmax-valid-hole-fits=N or -fno-max-valid-hole-fits)","replace _ with curry _","replace _ with ($) _","replace _ with ($!) _","replace _ with const _","replace _ with flip _","replace _ with id _"]
      filling infix type hole uses infix operator:          FAIL (1.69s)
        test/exe/Main.hs:5547:
        Found no matching actions for "replace _ with (+)" in ["replace _ with test","replace _ with (-)","replace _ with (*)","replace _ with asTypeOf","replace _ with const","replace _ with pure","replace _ with (Some hole fits suppressed; use -fmax-valid-hole-fits=N or -fno-max-valid-hole-fits)","replace _ with curry _","replace _ with ($) _","replace _ with ($!) _","replace _ with const _","replace _ with flip _","replace _ with id _"]
  Unit
    timestamps have millisecond resolution:                 FAIL (0.02s)
      test/exe/Main.hs:5717:
      Timestamps do not have millisecond resolution: 10000us
```

tests directly related with changes continue failing

**Comment by expipiplus1** [comment]
That's good news but my gripe was that when there are too many to choose
from at a glance then one may as well just type it. The speed at which they
appear wasn't really an issue for me here.

**Comment by Anton-Latukha** [comment]
Hole fitting solving would always be costly for the foreseeable futuristic future.

I consider 10 as a sane default (which would terminate mostly in the sane time & not clutter much, which I would prefer).

The option would be ideal, but it is an unrequested feature in a community project - if some style/project would need hundreds of hole fits - they would be able to arrive and implement it.

But lets not bikeshed - don't let the ideal be the enemy of the good.

---

@expipiplus1 would you be kind to push it over the last bump?

**Comment by expipiplus1** [comment]
Sorry, I won't be able to get to this in a timely manner. Happy for someone else to take over though.

**Comment by Anton-Latukha** [comment]
Yep. Looked into commits.

Rebasing for this diff was a bit excessive :rofl: 

---

Ok then somebody, me or others would finish this.

**Comment by dnikolovv** [comment]
Hey all.

What's left for this one to get in?

**Comment by michaelpj** [comment]
Superseded.

**Pull request closed without merging**